internal/controller/nnf_lustre_mgt_controller.go

```diff
@@ -0,0 +1,393 @@
+)
+
+type ControllerType string
+
+const (
+	ControllerWorker ControllerType = "Worker"
+	ControllerRabbit ControllerType = "Rabbit"
+	ControllerTest   ControllerType = "Test"
+)
+
+const (
+	// finalizerNnfLustreMGT defines the key used in identifying the
+	// storage object as being owned by this NNF Storage Reconciler. This
+	// prevents the system from deleting the custom resource until the
+	// reconciler has finished using the resource.
+	finalizerNnfLustreMGT = "nnf.cray.hpe.com/nnf_lustre_mgt"
+)
+
+// NnfLustreMGTReconciler contains the elements needed during reconciliation for NnfLustreMGT
+type NnfLustreMGTReconciler struct {
+	client.Client
+	Log          logr.Logger
+	Scheme       *kruntime.Scheme
+	ChildObjects []dwsv1alpha2.ObjectList
+
+	ControllerType ControllerType
+}
+
+//+kubebuilder:rbac:groups=nnf.cray.hpe.com,resources=nnflustremgts,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=nnf.cray.hpe.com,resources=nnflustremgts/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=nnf.cray.hpe.com,resources=nnflustremgts/finalizers,verbs=update
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=get;list;watch;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.2/pkg/reconcile
+func (r *NnfLustreMGTReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	log := r.Log.WithValues("NnfLustreMGT", req.NamespacedName)
+
+	metrics.NnfLustreMGTReconcilesTotal.Inc()
+
+	nnfLustreMgt := &nnfv1alpha1.NnfLustreMGT{}
+	if err := r.Get(ctx, req.NamespacedName, nnfLustreMgt); err != nil {
+		// ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	statusUpdater := updater.NewStatusUpdater[*nnfv1alpha1.NnfLustreMGTStatus](nnfLustreMgt)
+	defer func() { err = statusUpdater.CloseWithStatusUpdate(ctx, r.Client.Status(), err) }()
+	defer func() { nnfLustreMgt.Status.SetResourceErrorAndLog(err, log) }()
+
+	// Check if the object is being deleted.
+	if !nnfLustreMgt.GetDeletionTimestamp().IsZero() {
+		if !controllerutil.ContainsFinalizer(nnfLustreMgt, finalizerNnfLustreMGT) {
+			return ctrl.Result{}, nil
+		}
+
+		controllerutil.RemoveFinalizer(nnfLustreMgt, finalizerNnfLustreMGT)
+		if err := r.Update(ctx, nnfLustreMgt); err != nil {
+			if !apierrors.IsConflict(err) {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Add the finalizer
+	if !controllerutil.ContainsFinalizer(nnfLustreMgt, finalizerNnfLustreMGT) {
+		controllerutil.AddFinalizer(nnfLustreMgt, finalizerNnfLustreMGT)
```

Comment:
Maybe we don't have a reason to put a finalizer on this resource?

Reply:
I'm planning on adding in some logic on what to do if someone tries to remove the NnfLustreMGT resource while there are still claims on it, so I'd like to leave this in for now.